crates/iota-json-rpc-api/src/indexer.rs

```diff
@@ -45,7 +45,7 @@ pub trait IndexerApi {
 
     /// Return list of transactions for a specified query criteria.
     #[rustfmt::skip]
-    #[method(name = "queryTransactionBlocks")]
+    #[method(name = "queryTransactionBlocks", version <= "1.0.0")]
```

Comment:
We have to account for the largest release version prior to the target develop. E.g.
```suggestion
    #[method(name = "queryTransactionBlocks", version <= "1.2.10")]
```

Reply:
So considering the latest stable release it should be 1.2.3.
I tested i locally with curl and if no header is provided, it will target v2, so also the new filter is available.
If a header is provided, as expected it will target V1 if the provided version is < 1.2.3.

I also looked into the failing rpc-tests as they can't match the endpoint. I tried to inject a header, however this didn't help. I checked the raw request body that is received on the server and they use the correct internal method name, but it still somehow fail to match. When replacing the `client.query_transaction_blocks()` usages with `client.query_transaction_blocks_v2` the test passes.

Do you maybe know what I'm missing here? I assume they should at least work when a header is provided?

Reply:
> So considering the latest stable release it should be 1.2.3. I tested it locally with curl and if no header is provided, it will target v2, so also the new filter is available. If a header is provided, as expected it will target V1 if the provided version is <= 1.2.3.

The latest is `1.2.3` but the current develop is `1.3.0`. So we would like to catch any patches `1.2.x` until this new feature lands. And while we don't have a `< 1.3.0` which would be what we wanted here, we can leave some room by increasing the patch number. Hence the `1.2.10` that leaves room for additional patches to `1.2.3`.

> 
> I also looked into the failing rpc-tests as they can't match the endpoint. I tried to inject a header, however this didn't help. I checked the raw request body that is received on the server and they use the correct internal method name, but it still somehow fail to match. When replacing the `client.query_transaction_blocks()` usages with `client.query_transaction_blocks_v2` the test passes.
> 
>
> Do you maybe know what I'm missing here? I assume they should at least work when a header is provided?


Aren't the tests using the Rust API? It makes total sense that you should pass the `_v2` method to make use of the new filter. The version used above and the headers are relevant to http clients like `curl`.

Reply:
Thanks for the further clarification on the selected version.

> Aren't the tests using the Rust API? It makes total sense that you should pass the _v2 method to make use of the new filter. The version used above and the headers are relevant to http clients like curl.

Yes, the tests are using the Rust API. All clients that make use of the Rust API are directly trying to call the old method and will fail, with or without header.

Can be reproduced when adding following code to this [routing test](https://github.com/iotaledger/iota/blob/4d9b84037eafcebba24c5e8b83add60aba84f786/crates/iota-json-rpc-tests/tests/routing_tests.rs#L24): 
```rust
    let mut versioned_header = HeaderMap::new();
    versioned_header.insert(
        CLIENT_TARGET_API_VERSION_HEADER,
        HeaderValue::from_static("1.5"),
    );

    let client_with_new_header = HttpClientBuilder::default()
        .set_headers(versioned_header)
        .build(&url)
        .unwrap();

    let response = client_with_new_header
        .request(
            "test_foo_1_5",
            rpc_params!("old version expect string as input"),
        )
        .await;

    assert!(response.is_err())
```

Calling directly an old method (test_foo_1_5) with or [without](https://github.com/iotaledger/iota/blob/4d9b84037eafcebba24c5e8b83add60aba84f786/crates/iota-json-rpc-tests/tests/routing_tests.rs#L41) headers doesn't work.

So this is expected I assume. For the SDK it's not a problem as we can update the internal Rust API to v2.
For the tests we can update to v2 too.

Reply:
> Removing that branch makes the tests work, however we are no longer protected about unsupported direct function calls. A task for the future might be to revisit this routing logic and maybe reject on specific methods only.

There is indeed something off with that branch, but I would expect us to support direct calls actually. This seems to be a side effect of tweaking with the RPC macro and associating the Rust client method to the versioned variant, which is hidden by the spec and hence by other clients.

If direct access is not supported then the Rust API is no longer backward compatible.